napari/plugins/tests/test_manager.py

```diff
@@ -0,0 +1,60 @@
+import importlib
+import os
+import sys
+
+import pytest
+
+from napari.plugins import NapariPluginManager
+
+# from napari.plugins.manager import load_modules_by_prefix, validate_hookimpls
+
+PATH = os.path.join(os.path.dirname(__file__), 'fixtures')
+
+
+def setup_module():
+    """adding plugin test fixtures to path for autodiscovery"""
+    sys.path.append(PATH)
+
+
+def teardown_module():
+    sys.path.pop(sys.path.index(PATH))
+
+
+@pytest.fixture
+def pm():
+    pm = NapariPluginManager()
+    return pm
+
+
+def test_naming_convention_discovery(pm):
+    """make sure loading by naming convention works, and doesn't crash on
+    invalid plugins.
+
+    Note: loading by setuptools entry_point is tested in pluggy and is not
+    tested here.
+    """
+    assert 'napari_test_plugin' in pm._name2plugin
+    assert 'napari_bad_plugin' in pm._name2plugin
+
+    # napari_invalid_plugin has an invalid hookimpl, and will not get loaded
+    assert 'napari_invalid_plugin' not in pm._name2plugin
```

Comment:
Can you walk me through how these plugins end up in `pm._name2plugin`? When does the autodiscovery happen in this file?

Reply:
I'm thinking it happens automagically when PluginManager() gets instantiated, in which case, it seems too magical to me: I think there should be, at a minimum, a keyword argument to `__init__` that turns on the autodiscovery, so that at least, when there is calling code, it's *obvious* on reading it that it's doing that.

Reply:
Another idea: allow a path arg to the module discovery that does the adding and removing of the folder to sys.path. This would make it easy for people to maintain a "napari plugins" folder on their machine that they want to run...

Reply:
yeah that's not clear.  autodiscovery happens in the `pm` pytest fixture and yes, it happens during instantiation.  terrible :)  There _is_ a kwarg to init that turns on autodiscovery... but I should have added `autodiscover=True` to the fixture to make it more explicit. fixed.

Reply:
> Another idea: allow a path arg to the module discovery that does the adding and removing of the folder to sys.path. This would make it easy for people to maintain a "napari plugins" folder on their machine that they want to run...

ooo.  I _like_ that.  I was definitely thinking about adding something like that later (will be important for bundled apps), and that's as good of a place as any.

Reply:
added path argument to `discover()`.  going to resolve this conversation and leave the "whether to autodiscover or not" conversation for the other thread